Copy Translations

IsaiahPaget opened the pull request:
Added the ability to 'copy from' one language to another, this will aid users workflow because they will be able to see the structure of what they wrote in one language and make tweaks to it inline rather than manually copying or maybe even trying to remember what they wrote.

The original [issue(90)](https://github.com/wintercms/wn-translate-plugin/issues/90) was just about the repeater, but this PR includes this functionality for every ML formwidget.

I also updated the readme for some clearer examples of accessing json attributes on models.

IsaiahPaget commented on README.md:
```diff
@@ -189,6 +189,17 @@ class User extends Model
+{
+    public $implement = ['Winter.Translate.Behaviors.TranslatableModel'];
+
+    public $translatable = ['foo[bar][baz]'];
```
This could just be me, but when I was trying to figure this out my intuition was to use dot notation. So do with that what you will. I think it would make sense to me.

mjauvin replied:
> @mjauvin what are your thoughts on supporting dot notation here? I.e. `'foo.bar.baz'`?

Sure, why not.

AIC-BV commented:
There's something wrong at the mediafinder
![image](https://github.com/user-attachments/assets/21642ec2-bd48-4303-8433-732e572d6330)

AIC-BV commented:
The mediafinder field has wrong popup offset
![image](https://github.com/user-attachments/assets/cda6c90d-dd73-4209-8c8d-d7b6430d540b)

AIC-BV commented:
Perhaps some more padding on the textarea's so that the text is more/better visible
![image](https://github.com/user-attachments/assets/ddf14ec9-5125-4333-8735-836e092cb227)

IsaiahPaget commented:
@AIC-BV I fixed the overlap and the wrong dropdown positioning, do you wanna have another try at this branch see if it's working as expected? :)

IsaiahPaget commented:
> > @AIC-BV I fixed the overlap and the wrong dropdown positioning, do you wanna have another try at this branch see if it's working as expected? :)
> 
> Position is fixed! ![image](https://private-user-images.githubusercontent.com/89913092/463069460-d821366b-cd72-4d49-b611-1d1290a354ae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE5MDkyNzksIm5iZiI6MTc1MTkwODk3OSwicGF0aCI6Ii84OTkxMzA5Mi80NjMwNjk0NjAtZDgyMTM2NmItY2Q3Mi00ZDQ5LWI2MTEtMWQxMjkwYTM1NGFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA3VDE3MjI1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNWZlNjcxOTNjNjcyOTc1MDFmMWVmZmRiMTJjMWYyZTJjYTU1MWM1YTA2ZDMyODlhMjIyM2I1NmQ0MDc2M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.E47WszdIg9qnKN3W-5WA0YI-5J7jpcp00t-4tNdsA-E)
> 
> But the overlap is still there? Would say it would need atleast 4.5rem padding-right
> 
> ```
> .field-multilingual.field-multilingual-textarea textarea {
>     padding-right: 4.75rem;
> }
> ```

Ohh I am sorry I see what you mean, I was thinking you were just talking about the rich editor as per your previous screenshot, so thats what I changed, did you try that? The text area doesn't show the translate stuff when you are typing in it right, so I figured that was ok to leave the same. @LukeTowers what do you think? We could use the same button styling that is on the repeater and the editors on the textarea?

mjauvin commented:
@AIC-BV I have no objection, but I think we tried this before and it created issues.

AIC-BV commented:
For the auto-translate with ChatGPT...:

An extra feature is to be able to add a prompt in a textarea before submitting.

A simple/stupid example would be: translate "hallo" to "hey" instead of "hello"
This allows the user to manipulate certain translations in case the auto translation did something he/she doesn't like. We have many job-specific words which might get a wrong translation, or we call it different than an auto-translation would do

LukeTowers commented on traits/mlcontrol/partials/_locale_copy.htm:
```diff
@@ -0,0 +1,21 @@
+                tabindex="-1"
+                data-copy-locale="<?= $code ?>"
+                href="javascript:;">
+                Copy from <?= $name ?>
```
```suggestion
                <?= e(trans('winter.translate::lang.locale.copy_from', ['locale' => $name])); ?>
```
then add the `copy_from` key to `lang/en/lang.php` under the `locale` key with the value `Copy from :locale`

IsaiahPaget commented on formwidgets/mlmarkdowneditor/partials/_mlmarkdowneditor.htm:
ohh whoops, thanks, I have a shortcut to make files executable but it's too easy to press lol, I gotta update that.

mjauvin commented:
I prefer this as an option disabled by default.

LukeTowers commented:
@mjauvin could you elaborate on that? What are your objections to having it enabled always and / or by default?

mjauvin commented:
> @mjauvin could you elaborate on that? What are your objections to having it enabled always and / or by default?

In my opinion, if the functionality is not going to be used for specific people, it will clutter the UI more than it already does.

I don't really mind if it's on by default (as long as it can easily be changed programatically), but I really want this as an optional feature.

LukeTowers commented on formwidgets/mlmarkdowneditor/assets/js/mlmarkdowneditor.js:
```diff
@@ -37,25 +39,28 @@
         this.$textarea.on('changeContent.oc.markdowneditor', this.proxy(this.onChangeContent))
 
         this.updateLayout()
+        this.initInputEvents()
```
@IsaiahPaget could you make sure these events are disposed of in the dispose function as well?

LukeTowers commented on formwidgets/mlmarkdowneditor/assets/js/mlmarkdowneditor.js:
```diff
@@ -76,40 +86,64 @@
+        this.updateLayout()
+    }
+
+
```
```suggestion
```

mjauvin commented:
@IsaiahPaget using this PR in version 2.3.1, the placement for the language selector / copyfrom button is all over the place... what am I missing?

Also, there should have been an entry added to the config to show how to disable the copyfrom... I see it in the code, but no mention in the config file.

mjauvin commented:
Somehow, I had to force my browser to clear its cache to get the new multilingual CSS... not sure why as the `addCss()` call uses the plugin name, which should have used the new plugin version (which should have forced the new css file to get fetched)... 

All working fine now.

arvislacis commented:
I have come accross some issues (https://github.com/wintercms/wn-translate-plugin/issues/108) with accessing form fields because copy translations drop-down even in it's collapsed state is overflowing the form elements.

As I wrote in my issue, I am not sure why the `height: 100%;` property is needed at all, or it should be replaced with property which takes up only the real height needed for drop-down content and/or should be applied only when the actual drop-down element has been opened/shown.
